src/pykeen/nn/node_piece/anchor_search.py

```diff
@@ -222,6 +224,175 @@ def __call__(self, edge_index: numpy.ndarray, anchors: numpy.ndarray, k: int) ->
+        """
+        # infer shape
+        num_entities = edge_index.max().item() + 1
+        edge_index = torch.as_tensor(edge_index, dtype=torch.long)
+
+        # symmetric + self-loops
+        edge_list = torch.cat(
+            [edge_index, edge_index.flip(0), torch.arange(num_entities).unsqueeze(0).repeat(2, 1)], dim=-1
+        ).unique(
+            dim=1
+        )  # unique for deduplicating repeated edges
+
+        return edge_list
+
+    @staticmethod
+    def bfs(
+        anchors: numpy.ndarray,
+        edge_list: torch.tensor,
+        max_iter: int,
+        k: int,
+        device: torch.device,
+    ) -> numpy.ndarray:
+        """
+        Determine the candidate pool using breadth-first search.
+
+        :param anchors: shape: (a,)
+            the anchor node IDs
+        :param edge_list: shape: (2, n)
+            the edge list with symmetric edges and self-loops
+        :param max_iter:
+            the maximum number of hops to consider
+        :param k:
+            the minimum number of anchor nodes to reach
+        :param device:
+            the device on which the calculations are done
+
+        :return: shape: (n, a)
+            a boolean array indicating whether anchor $j$ is in the set of $k$ closest anchors for node $i$
+
+        :raises ImportError:
+            If :mod:`torch_sparse` is not installed
+        """
+        try:
+            import torch_sparse
+        except ImportError as err:
+            raise ImportError("Requires `torch_sparse` to be installed.") from err
+
+        num_entities = edge_list.max().item() + 1
+        # for each entity, determine anchor pool by BFS
+        num_anchors = len(anchors)
+
+        anchors = torch.tensor(anchors, dtype=torch.long, device=device)
+
+        # an array storing whether node i is reachable by anchor j
+        reachable = torch.zeros((num_entities, num_anchors), dtype=torch.bool, device=device)
+        reachable[anchors] = torch.eye(num_anchors, dtype=torch.bool, device=device)
+
+        # an array indicating whether a node is closed, i.e., has found at least $k$ anchors
+        final = torch.zeros((num_entities,), dtype=torch.bool, device=device)
+
+        # the output that track the distance to each found anchor
+        # dtype is unsigned int 8 bit, so we initialize the maximum distance to 255 (or max default)
+        # TODO int8 takes 8x more space than just a binary matrix, decide if we want to keep it
+        dtype = torch.uint8
+        pool = torch.zeros((num_entities, num_anchors), dtype=dtype, device=device).fill_(torch.iinfo(dtype).max)
+        # initial anchors are 0-hop away from themselves
+        pool[anchors, torch.arange(len(anchors), dtype=torch.long, device=device)] = 0
+
+        edge_list = edge_list.to(device)
+        values = torch.ones_like(edge_list[0], dtype=torch.bool, device=device)
+
+        # TODO: take all (q-1) hop neighbors before selecting from q-hop
```

Comment:
can you write some more context for what to do to fix this in the future and why it's necessary?

Reply:
good catch, we actually don't need this TODO anymore - the `pool` matrix tracks all anchors and their distances, so it's done in a desired way